src/storage/schema.rs

```diff
@@ -0,0 +1,265 @@
+            patch,
+            revision,
+        }
+    }
+}
+
+impl fmt::Display for SchemaVersion {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}.{}.{}-{}",
+            self.major, self.minor, self.patch, self.revision
+        )
+    }
+}
+
+pub fn current_schema_version(tx: &mut Transaction) -> Result<Option<SchemaVersion>> {
+    let sql = "
+SELECT *
+FROM version";
+
+    fn convert_row(row: super::Row) -> Result<SchemaVersion> {
+        let revision = if row.column_count() > 3 {
+            row.get(3)?
+        } else {
+            // Might be prior to schema 0.7.0-1 where there was no revision column
+            0
+        };
+        Ok(SchemaVersion {
+            major: row.get(0)?,
+            minor: row.get(1)?,
+            patch: row.get(2)?,
+            revision,
+        })
+    }
+
+    let res = tx.query_single(sql, convert_row);
+
+    // Ignore errors
+    // TODO: this is similar to the Go implementation, but it would be better to ignore only the
+    // case where the table does not exist.
+    Ok(res.ok().flatten())
+}
+
+pub fn insert_schema_version(tx: &mut Transaction, version: &SchemaVersion) -> Result<()> {
+    let sql = "
+INSERT INTO version (major, minor, patch, revision)
+VALUES (?, ?, ?, ?)";
+
+    let params = rusqlite::params![
+        version.major,
+        version.minor,
+        version.patch,
+        version.revision
+    ];
+    match tx.execute_params(sql, params) {
+        // TODO: is all this error handling really necessary?
+        Ok(1) => Ok(()),
+        Ok(_) => {
+            Err("Version could not be inserted: expected exactly one row to be affected".into())
+        }
+        Err(e) => Err(e.chain_err(|| "Could not insert schema version")),
+    }
+}
+
+pub fn update_schema_version(tx: &mut Transaction, version: &SchemaVersion) -> Result<()> {
+    let sql = "
+UPDATE version SET major = ?, minor = ?, patch = ?, revision = ?";
+
+    let params = rusqlite::params![
+        version.major,
+        version.minor,
+        version.patch,
+        version.revision
+    ];
+    match tx.execute_params(sql, params) {
+        // TODO: is all this error handling really necessary?
+        Ok(1) => Ok(()),
+        Ok(_) => {
+            Err("Version could not be updated: expected exactly one row to be affected".into())
+        }
+        Err(e) => Err(e.chain_err(|| "Could not update schema version")),
+    }
+}
+
+pub fn create_schema(tx: &mut Transaction) -> Result<()> {
+    create_tag_table(tx)?;
+    create_value_table(tx)?;
+    create_file_table(tx)?;
+    create_file_tag_table(tx)?;
+    create_implication_table(tx)?;
+    create_query_table(tx)?;
+    create_setting_table(tx)?;
+    create_version_table(tx)?;
+    insert_schema_version(tx, &LATEST_SCHEMA_VERSION)?;
+    Ok(())
+}
+
+fn create_tag_table(tx: &mut Transaction) -> Result<()> {
+    tx.execute(
+        "
+CREATE TABLE IF NOT EXISTS tag (
```

Comment:
I find this ugly, but `rustfmt` insists on having the double quote on its own line...